Modix.Services/Moderation/ModerationService.cs

```diff
@@ -990,55 +1015,65 @@ await DeletedMessageBatchRepository.CreateAsync(new DeletedMessageBatchCreationD
+            InfractionSummary? infraction,
+            string? reason = null,
             bool isAutoRescind = false)
         {
-            RequestOptions GetRequestOptions() =>
+            RequestOptions? GetRequestOptions() =>
                 string.IsNullOrEmpty(reason) ? null : new RequestOptions {AuditLogReason = reason};
 
-            if (infraction == null)
-                throw new InvalidOperationException("Infraction does not exist");
-
-            if (!isAutoRescind)
-                await RequireSubjectRankLowerThanModeratorRankAsync(infraction.GuildId,
-                    AuthorizationService.CurrentUserId.Value, infraction.Subject.Id);
-
-            await InfractionRepository.TryRescindAsync(infraction.Id, AuthorizationService.CurrentUserId.Value, reason);
-
-            var guild = await DiscordClient.GetGuildAsync(infraction.GuildId);
+            var guild = await DiscordClient.GetGuildAsync(guildId);
 
-            switch (infraction.Type)
+            switch (type)
             {
                 case InfractionType.Mute:
-                    if (!await UserService.GuildUserExistsAsync(guild.Id, infraction.Subject.Id))
+                    if (!await UserService.GuildUserExistsAsync(guild.Id, subjectId))
                     {
                         Log.Information(
                             "Attempted to remove the mute role from {0} ({1}), but they were not in the server.",
-                            infraction.Subject.GetFullUsername(),
-                            infraction.Subject.Id);
+                            infraction?.Subject.GetFullUsername() ?? "Unknown user",
+                            subjectId);
                         break;
                     }
 
-                    var subject = await UserService.GetGuildUserAsync(guild.Id, infraction.Subject.Id);
+                    var subject = await UserService.GetGuildUserAsync(guild.Id, subjectId);
                     await subject.RemoveRoleAsync(await GetDesignatedMuteRoleAsync(guild), GetRequestOptions());
                     break;
 
                 case InfractionType.Ban:
-                    await guild.RemoveBanAsync(infraction.Subject.Id, GetRequestOptions());
+                    await guild.RemoveBanAsync(subjectId, GetRequestOptions());
                     break;
 
                 default:
-                    throw new InvalidOperationException($"{infraction.Type} infractions cannot be rescinded.");
+                    throw new InvalidOperationException($"{type} infractions cannot be rescinded.");
+            }
+
+            if (infraction != null)
+            {
+                if (!isAutoRescind)
+                {
+                    await RequireSubjectRankLowerThanModeratorRankAsync(guildId,
+                        AuthorizationService.CurrentUserId.Value, subjectId);
+                }
+
+                await InfractionRepository.TryRescindAsync(infraction.Id, AuthorizationService.CurrentUserId.Value, reason);
```

Comment:
+ null-forgiving operators to suppress the nullable warnings

```suggestion
                if (!isAutoRescind)
                {
                    await RequireSubjectRankLowerThanModeratorRankAsync(guildId,
                        AuthorizationService.CurrentUserId!.Value, subjectId);
                }

                await InfractionRepository.TryRescindAsync(infraction.Id, AuthorizationService.CurrentUserId!.Value, reason);
```